### ISSUE TITLE
reverse shell

### ISSUE THREAD

**Issue by kodeine** [vesche/basicRAT#17]
I see a PR which has a reverse shell but its not implemented in the master branch?
Can you please tell me when reverse shell is expected?

**Comment by vesche** [comment]
I'm assuming you mean an interactive shell? I've prototyped this quite a bit. The PR implemented a pseudo reverse shell that would interpret directory changes and use the `os` library for the current directory. This isn't a true reverse shell and could causes problems with things like relative paths in command line arguments. I'm instead prototyping a solution that would actually send an interactive `cmd.exe` or `bash` over the wire, similar to how netcat works. Not sure when this will be implemented, but I'll leave this open until it is.

**Comment by reznok** [comment]
It handled relative paths just fine I believe. Would need to be tested more, but the few things I tried to do with it worked. You're right of course that it's not a "true shell", but it was functional. I'd be interested to hear why you want to go the netcat-esque method. I'm assuming I'm overlooking a benefit you get with it :stuck_out_tongue:

**Comment by deadPix3l** [comment]
one benefit that it has is the ability to maintain environment variables, as well as aliases, and other settings that an interactive session has (eg. .bash_profile). also if alternative shells are available (fish, zsh, etc) they can provide additional features beyond the os library.

**Comment by kodeine** [comment]
@deadPix3l @vesche 
I am thinking why this rat was in python, i tried using pyinstaller as onefile and the filesize is > 5mb. Or is there a way to make it more smaller, i tried going to GOlang way too but same, its not as big as python one file but still > 3mb.

**Comment by vesche** [comment]
@reznok What I mean about relative paths is say you ran the command `cp /path/to/file long/relative/path/` the pseudo reverse shell won't be able to handle relative paths as arbitrary command line arguments. @deadPix3l outlined some other concerns.

@kodeine What? This RAT is written in Python, not Golang... No matter what your executable is going to be a couple megabytes. You can exclude libraries from the stdlib, by using a PyInstaller spec file. [See the PyInstaller docs about spec file operation](https://pythonhosted.org/PyInstaller/spec-files.html#spec-file-operation), just list them in the excludes section.

**Comment by vesche** [comment]
As an update to this thread, my progress on what I outlined above https://github.com/vesche/basicRAT/issues/17#issuecomment-285953438 culminated in a stack overflow post you can read [here](http://stackoverflow.com/questions/42767321/python-true-interactive-remote-reverse-shell). It has gone unanswered, so I'm still trying to figure this out. I might end up making a disgusting abstraction of a shell like I didn't want to do. We'll see.

**Comment by reznok** [comment]
So you're essentially trying to replicate the functionality of SSHing into the box, but as a reverse shell? 
To address @deadPix3l's points, yes those would be great, but should that be stored client side? I would say persistent data like that should be stored on the server side and processed before sending commands (like aliases, etc.) unless you want the rat leaving artifacts all over the place. I guess it comes down to your overall vision for this: a stealth solution or a fully functional shell solution.

A possible idea though would be to use something like Paramiko to establish an SSH server on the client, then from client side SSH to server and build a reverse tunnel that could be used to hit the client's SSH server. 
So the logic:
RATClient establishes SSH server
RATClient establishes SSH connection to SSH Server on RATServer with reverse tunnel
RATServer establishes SSH connection to SSH server on RATClient through tunnel

That would only be needed if you need to skate around firewalls. If you were able to directly connect to the client, then there's no need for the reverse tunnel.

**Comment by deadPix3l** [comment]
@reznok yes, in most cases data like that should be kept server side but sometimes that is not possible or the current tools being used on the the client don't work that way. Two examples: 1. the client side has certain environment variables set, and their values are desired. these variables cannot be accessed through the pseudo shells you provided. only a real shell will do. 2. a command that was run on the client sets a variable. this is not an issue because once the shell is ended that value is destroyed, leaving no trace. it is honestly the best way to store small amounts of data on the client if needed imo. in a pseudoshell, these variables are destroyed immediately. and therefore that tools is unusable. these, along with many others, are reasons why a true shell is desirable.

as for the other ideas (we appreciate your understanding, and we do appreciate ideas for how to make this work):

we looked at paramiko and we dont believe it is what we are looking for, but we will look into it further.
we also looked at a few other technologies and are determined to figure this out.

Thank you

**Comment by n1nj4sec** [comment]
I guess you are just looking for a Popen on /bin/sh or cmd.exe to write your commands on stdin ... https://github.com/n1nj4sec/pupy/blob/unstable/pupy/packages/all/interactive_shell.py
